In [1]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

/var/folders/x4/l7tc61hd00n1s8vbqfwjq7k40000gn/T/ipykernel_77005/4005703837.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


---

In [ ]:
# Confirm the environment to run program
# !conda env list

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv("/Users/damiensmac/Desktop/Credit_Risk_Chal/credit-risk-classification/Credit_Risk/Starter_Code/Resources/lending_data.csv")

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df["loan_status"]

# Separate the X variable, the features
X = df.drop(columns="loan_status")


In [5]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [6]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)


# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
test_accuracy = balanced_accuracy_score(y_test, testing_predictions)

print(f"The balanced accuracy is {test_accuracy}")

The balanced accuracy is 0.9520479254722232


In [12]:
# Generate a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)

print(f"The confusion matrix is: {test_matrix}")

The confusion matrix is: [[18663   102]
 [   56   563]]


In [13]:
# Print the classification report for the model
test_report = classification_report(y_test, testing_predictions)

print(f"The Classification report for the testing data is :\n{test_report}")

The Classification report for the testing data is :
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The LRM's 'Precision' in predicting (healthy loans) is 1.0 with a 'Recall' of .99 and 'F1' of 1.0 suggesting that the model is successful in predicting those loans. However, regarding (high-risk loans) the model only scored .85(precision) with an 'F1' score of .88

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [ ]:
# import sklearn
# print("Scikit-learn (sklearn) version:", sklearn.__version__)
# pip install scikit-learn==1.0.2
# pip install imbalanced-learn

In [14]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
x_res, y_res = random_oversampler_model.fit_resample(X_train, y_train)

In [15]:
x_res.value_counts()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  derogatory_marks  total_debt
18500.0    10.976         83900            0.642431        12               2                 53900         209
17500.0    10.552         79900            0.624531        11               2                 49900         181
17400.0    10.510         79500            0.622642        11               2                 49500         157
18400.0    10.939         83500            0.640719        12               2                 53500         153
19300.0    11.329         87200            0.655963        12               2                 57200         140
                                                                                                           ... 
12900.0    8.594          61500            0.512195        7                1                 31500           1
           8.598          61500            0.512195        7                1                 31500           1

In [16]:
y_res.value_counts()

loan_status
0    56271
1    56271
Name: count, dtype: int64

In [17]:
from collections import Counter

In [18]:
# Count the distinct values of the resampled labels data
label_counts = Counter(y_res)
print(label_counts)

Counter({0: 56271, 1: 56271})


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [19]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
classifier.fit(x_res, y_res)

# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({'Prediction': predictions, "Actual": y_test})



Prediction  Actual
60914           0       0
36843           0       0
1966            0       0
70137           0       0
27237           0       0
...           ...     ...
45639           0       0
11301           0       0
51614           0       0
4598            0       0
2793            0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [20]:
# Print the balanced_accuracy score of the model 
ba_score = balanced_accuracy_score(y_test, predictions)
print(ba_score)

0.9936781215845847


In [21]:
# Generate a confusion matrix for the model
prediction_matrix = confusion_matrix(y_test, predictions)
print(prediction_matrix)

[[18649   116]
 [    4   615]]


In [22]:
# Print the classification report for the model
cr = classification_report(y_test, predictions)
print(cr)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** This model performed best in predicting "Healthy Loans"  as represented with it's precision, recall and F1-Score all performing near optimal levels. However, in terms of "High Risk Loans", it performed slightly weaker as represented in it's precision of only .84, suggesting that further focus is required to counteract oversampling.